### DISCUSSION TITLE
Make email templates translatable

### DISCUSSION THREAD

**Discussion by rijkvanzanten** [directus/directus#8239] [Feature Requests]
Currently, the email templates for reset-password and invite-user are hardcoded in English. We should figure out a way to use translation strings within these email templates in order to be able to send the email itself in the user's configured language.

We should also consider how this would work with custom email templates in mind.

**Comment by joselcvarela** [comment]
What should the language be of the invite-user email? Since the user is not yet on the platform, we can only rely on the inviter language or provide a way to specify the language for the email.

**Reply by rijkvanzanten** [reply]
I think it should be based on the project's default language that @christianrr added support for in https://github.com/directus/directus/pull/8196 🙂

**Reply by joselcvarela** [reply]
Nice. We could also think that we might want to invite users in other languages, but seems an edge use case.

**Reply by rijkvanzanten** [reply]
Yeah, that feels a little overkill to me

**Reply by ludi81** [reply]
In my opinion the forgot password function is more important than the invitation message.
But still, the invitation message or registration message should use the Accept-Language (http header) or if this is not set the default language of the project

**Reply by rijkvanzanten** [reply]
> should use the Accept-Language (http header)

I don't agree with this. The email is send through an API endpoint, which can be called from any environment, browser or otherwise. Relying on a http header coming from the browser can lead to pretty unexpected results.

**Reply by ludi81** [reply]
If you don't trust the Accept-Language you can use an own language parameter. We would like to choose the language of the messages (e.g) reset password). Only the client knows which language is currently used.

At the moment we send one email with multiple languages (e.g. reset password) .

**Reply by rijkvanzanten** [reply]
> Only the client knows which language is currently used.

Does it though? If you reset the password of a user, we know what the user's configured language is and can use that 👍🏻

**Reply by ludi81** [reply]
you are right. What's with the registration message?

**Reply by rijkvanzanten** [reply]
At that point, we don't know what the receiving user's preference is, and we don't want to rely on the current user's language. The safest option is to default to the project-default language (currently English, but configurable in the future)

**Reply by ludi81** [reply]
But the client knows when calling the API which language is currently used. Why could not send the client the expected language (e.g. parameter)? Thats the reason why I suggested the Accept-Language  header. But you are right, if the normal browser is used you may not rely on this header.

**Reply by joselcvarela** [reply]
@ludi81 you are relying on Accept-Language only for the case of the user clicks on the link for password-reset. This might be the only exception on all e-mails we send. On all other cases we need to rely on server side data, not on the client side, since we want to send e-mails when users are not currently online.

**Reply by w0kyj** [reply]
Couple of thoughts from my somewhat limited point of view:
- Add an option for an additional variable (language_code) or perhaps even a variable array that may get a variety of key values to be used in the liquid templates.
- Seems like we should be using liquid templates for all system emails instead of generating in code.  Much more flexible and transparent for my simple thinking :)  This also leaves it flexible to use:
    - Template naming conventions to differentiate translations, e.g. password-reset-en | password-reset-de | ...
    - Template conditional logic to differentiate translations, e.g. if language_code = en-EN then "english ..." elsif language_code = de-DE then "Deutsch ..." else "Default language ..."

**Reply by joselcvarela** [reply]
I think it can be easier and flexible.
Instead of expecting a liquid template, we could expect a JS file to export a function where the user could make his own queries to database and everything he wants which will then returns all the email options including the html, from, to and could also use bcc, cc and other options passed to nodemailer.

**Reply by rijkvanzanten** [reply]
@w0kyj 95% of the template will be the same between the different languages. Things like styling, headers, layout, button placement etc etc is all the same. It'll be fairly difficult to maintain that if every language has a full copy of all those shared things 🤔

**Reply by w0kyj** [reply]
If we did the JS file - then we could support things like this example:  https://github.com/harttle/liquidjs/tree/master/demo/nodejs - provides way to manage headers, footers, layouts templates in liquid, then use the db with translations to get data for the text, etc... ???

**Reply by rijkvanzanten** [reply]
That's a very similar approach as to what I mentioned before.. somewhere.. in some other discussion I think, anyway:

I was imagining we could register a custom "translate" filter in LiquidJS, which would then use a similar `t()` setup as we use in the app. This would allow the template to look something like:

```liquid
<p>{{ email_intro | translate }}</p>
```

With the actual strings being pulled from a centralized translations file that can be managed through https://locales.directus.io
```yaml
en-us:
  email_intro: Hi there!
nl-nl:
  email_intro: Hoi!
```

That way, we can make sure the HTML, styling, everything that isn't translated text stays centralized and easy to maintain 🙂

**Reply by joselcvarela** [reply]
@rijkvanzanten you could be right, but I was thinking about this https://github.com/directus/directus/discussions/4002
I think developers should be able to customize the emails to be sent. On the other hand, they actually can do it already by disabling the mail system of Directus and by implementing their own system with custom hooks. I have a mixed feelings about this 😵

**Reply by rijkvanzanten** [reply]
> @rijkvanzanten you could be right, but I was thinking about this https://github.com/directus/directus/discussions/4002
> 
> I think developers should be able to customize the emails to be sent. On the other hand, they actually can do it already by disabling the mail system of Directus and by implementing their own system with custom hooks. I have a mixed feelings about this 😵

We already load in custom email templates from the extensions directory. If you name it the same as the system templates, it'll use the custom one instead :) 

That part doesn't solve for translations though, which my approach above would add 👍🏻

**Reply by naskio** [reply]
Hello
I would suggest to add an env variable to set a default language for the whole project.
If the variable is set, the default language will be used as default for all the users.
Users can set a different language by updating the profile.
It will also be used for email templates, etc.
I would suggest also to add an env variable to limit the languages to certain languages (instead of supporting all the languages you can decide to choose only a subset).
This will necessarily include a way to add new translation keys and a way to specify the translations strings (the default language will be used as a fallback language if the translation for certain keys is missing).

**Reply by bereikme** [reply]
@rijkvanzanten What's the current status on email translations?

**Reply by rijkvanzanten** [reply]
> @rijkvanzanten What's the current status on email translations?

No new updates yet 👍🏻

**Reply by DamienMrtl** [reply]
![image](https://user-images.githubusercontent.com/7550391/172411844-9f887760-b41e-4d95-a0e5-9a7ae990d05c.png)
There should be an option to select the language for the invited user here, same as we can choose a role. I don't find this overkill at all.

Or maybe as an alternative, we could manually create an user and set only its email and language for example. And then on the user page we could have a button to send/re-send the invitation mail. So in this case the mail would be directly in the user language.

**Reply by benhaynes** [reply]
Good one! We'll have to wait until this feature request gets a bit higher to start in, though... with our limited resources we have to follow sponsored work and community upvotes. Still, will  be great to get to this in time!

**Reply by snorbik** [reply]
@benhaynes Any udpate about this? We're currently working on a multilanguage app having Directus as its BE, and English hardcoded reset mails are quite complicated as language used is not English in the app.

**Reply by benhaynes** [reply]
No updates, I'm afraid. The engineers are extremely busy at the moment, so we'll have to wait a bit for this one... or see if someone wants to sponsor its expedited development.

**Reply by snorbik** [reply]
Hi @benhaynes, we've been thinking about it again in our team. What we see as the best solution for the Directus team would be to add those strings that are used in invitation and password reset emails to the official language repo on Crowdin and then just use strings based on the user's set language. This wouldn't allow someone to change those text and "hack" the process and would give better internationalization to Directus apps everywhere. Do you think the approach would be possible?

**Reply by Pachat** [reply]
That is a point of view aimed to standardization and I understand it this way.

IMHO, the aim of a product as Directus is to give the maximum versatility upon solid implementations.

On my side, the content of the mail templates do not fit the needs of my application (and it is not a question of translation).
As I use only one language at the moment, I ended modifying the templates themselves directly.

My feeling is that `language templates` would be a better solution:
Ex.: `ca-ES.password-reset.liquid` used if it exists, otherwise `password-reset.liquid` as default.

**Reply by snorbik** [reply]
Our idea is the same as was presented here in a thread in September 2021.

> With the actual strings being pulled from a centralized translations file that can be managed through https://locales.directus.io

Was there any update on that @rijkvanzanten ? Or is there any way we can help?

**Reply by rijkvanzanten** [reply]
@snorbik No updates yet. We're prioritizing new feature requests [based on community interest](https://github.com/directus/directus/discussions/categories/ideas?discussions_q=category:Ideas+sort:top) and at the moment this isn't as highly ranked as other ideas.

While pulling the strings from [Crowdin](https://locales.directus.io) is most likely the way to go, it's not as trivial as just adding the strings to the existing config. The email template content is generated API-side, which currently doesn't have any translations support at all.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by w0kyj** [comment]
The current system emails:
- user invite
- password reset
- user registration

Possible Solutions:
- New `directus_emails` or similar system settings collection(s) with translations support to manage the system emails and any other emails the admin would like
- Allow for Crowdin translations of the system emails
- Add language selection to user invites that defaults to project language

**Comment by rbochenski** [comment]
Perhaps something like that would solve the problem:
https://github.com/directus/directus/pull/22452

**Comment by Pachat** [comment]
A quick and easy solution in case you want just _one_ language:
In docker-compose.yml add
```
      - /var/www/mydomain.com/node_modules/@directus/api/dist/services/users.js:/directus/node_modules/@directus/api/dist/services/users.js
      - /var/www/mydomain.com/node_modules/@directus/api/dist/services/mail/templates/base.liquid:/directus/node_modules/@directus/api/dist/services/mail/templates/base.liquid
      - /var/www/mydomain.com/node_modules/@directus/api/dist/services/mail/templates/password-reset.liquid:/directus/node_modules/@directus/api/dist/services/mail/templates/password-reset.liquid
      - ...
```
and put the external files adapted in your language at the left paths above and restart docker.
_Test and adapt in case of upgrades._

**Reply by ricardovigatti** [reply]
against any recommendedations of "do not touch" 3rd-party code, sadly we did that for a client since its the only solution to translate the email for they users.

**Comment by lucameusburger** [comment]
My quick and dirty hack for the moment

```
import { defineHook } from '@directus/extensions-sdk'

export default defineHook(({ filter }) => {

  filter('email.send', async (input: any) => {

    if (input.subject === 'Password Reset Request') {
      input.subject = 'Passwort festlegen'
    }

    return input
  })
})
```

**Reply by ricardovigatti** [reply]
It looks good. Definitely will try the next time.

**Reply by phazonoverload** [reply]
This is a fairly new event that we added in lieu of having translatable/configurable system emails. I'm glad you found it.

**Reply by jekuer** [reply]
However, there is no chance to get the language of the user here, right? At least not in the "invite case".

**Reply by phazonoverload** [reply]
Well a newly-invited user doesn't have any settings, so there won't be a 'user language' yet. That said, you can still intercept this email using the above approach (hooks or flows) and apply whatever logic you want.

**Reply by lucameusburger** [reply]
I guess for a newly-invited user you could still get the language from the req.headers accept-language. and for an existing user you can grab the accountability.user and get the language from there.

**Reply by jekuer** [reply]
No. You do not have access to the request or information about the inviting user in the hook (afaik).

**Reply by jekuer** [reply]
But I learned that in all those cases, the user is already created and can be identified via the input.to email address. Will share some code later.

**Reply by jekuer** [reply]
As promised: https://github.com/jekuer/directus-extension-system-email-i18n
First draft. Working fine on my side for my case. Feedback welcome!
